module/zfs/vdev_raidz.c

```diff
@@ -2936,6 +2936,74 @@ vdev_raidz_xlate(vdev_t *cvd, const range_seg64_t *in, range_seg64_t *res)
 	ASSERT3U(res->rs_end - res->rs_start, <=, in->rs_end - in->rs_start);
 }
 
+
+static void
+raidz_scratch_read_done(zio_t *zio)
+{
+	zio_nowait(zio_unique_parent(zio));
+}
+
+static void
+raidz_scratch_write_done(zio_t *zio)
+{
+	vdev_raidz_expand_t *vre = zio->io_private;
+
+	if (zio->io_error == 0)
+		vre->vre_scratch_devices++;
+
+	abd_free(zio->io_abd);
+
+	spa_config_exit(zio->io_spa, SCL_STATE, zio->io_spa);
+}
+
+static void
+raidz_scratch_object(spa_t *spa, vdev_raidz_expand_t *vre)
+{
+	(void)vre;
+
+	vdev_t *raidvd = vdev_lookup_top(spa, vre->vre_vdev_id);
+
+	for (int i = 0; i < raidvd->vdev_children - 1; i++) {
+		dmu_tx_t *tx =
+		    dmu_tx_create_dd(spa_get_dsl(spa)->dp_mos_dir);
+		VERIFY0(dmu_tx_assign(tx, TXG_WAIT));
+		uint64_t txg = dmu_tx_get_txg(tx);
+
+		spa_config_enter(spa, SCL_CONFIG, FTAG, RW_READER);
+		raidvd = vdev_lookup_top(spa, vre->vre_vdev_id);
+
+		int txgoff = dmu_tx_get_txg(tx) & TXG_MASK;
+
+		spa_config_enter(spa, SCL_STATE, spa, RW_READER);
+
+		zio_t *pio = spa->spa_txg_zio[txgoff];
+		abd_t *abd = abd_alloc_for_io(VDEV_BOOT_SIZE, B_FALSE);
+		zio_t *write_zio = zio_vdev_child_io(pio, NULL,
+		    raidvd->vdev_child[i],
+		    -VDEV_BOOT_SIZE,
+		    abd, VDEV_BOOT_SIZE,
+		    ZIO_TYPE_WRITE, ZIO_PRIORITY_REMOVAL,
+		    ZIO_FLAG_CANFAIL,
+		    raidz_scratch_write_done, vre);
+
+		zio_nowait(zio_vdev_child_io(write_zio, NULL,
+		    raidvd->vdev_child[i],
+		    0,
+		    abd, VDEV_BOOT_SIZE,
```

Comment:
When I considered boot sector as scratch object, the next calculations were used:
VDEV_BOOT_SIZE eq 3.5M (7ULL << 19) per child. Max number of childs is 255.
255*(7ULL << 19) = 935854080 bytes => max scratch object size

nchildren <<ashift in case of 4k ashift:
255 << 12 = 1044480 bytes

So, if I am correct it should be enough, of course excluding bootloader if it present, but I prefere do not accaunt it for now.
Later it will be needed to add logic which will detect the boot loader and check it size.
The simplest way will be to zero boot sector on pool creatrion phase and check non-zeroed data size to accaunt it.

Also, you mentioned "Copy first new_ncols^2 sectors to scratch object".
Could you please explain why new_ncols^2 and from other side "nchildren <<ashift". I cannot get it clear.

Reply:
Sure, I think the design doc talks about "Copy first new_ncols^2 sectors to scratch object".  This is the same as each device having "nchildren << ashift" bytes copied to it.  Since each device has nchildren sectors, and there are nchildren devices, we have nchildren * nchildren = nchildren^2 sectors copied total (across all devices).

The reason we need to copy this much data is so that after the scratch data, there is at least one whole stripe (aka one whole row, which is at most `old_nchildren << ashift` bytes) of not-yet-overwritten old data before the next old block to copy.  We need this because we read this old data (from just before the "next old block to copy" when there's a stripe that crosses the "next block to copy" boundary, `vre_offset_phys`.  This is implemented by the code in vdev_raidz_map_alloc_expanded() following this comment:
```
		 * If we are in the middle of a reflow, and any part of this
		 * row has not been copied, then use the old location of
		 * this row.
```